File: client/src/services/binanceMarketService.ts
import type { ChartCandle } from '@/types/candle';

const BINANCE_REST_BASE = 'https://api.binance.com';
/** Single raw stream: /ws/{streamName} */
const BINANCE_WS_SINGLE = 'wss://stream.binance.com:9443/ws';
/** Combined streams: /stream?streams=... (not under /ws) */
const BINANCE_WS_COMBINED = 'wss://stream.binance.com:9443';

export type BinanceKlineInterval = '1s' | '1m' | '5m' | '15m' | '1h';

export type BinanceConnectionStatus =
	| 'idle'
	| 'loading'
	| 'connected'
	| 'reconnecting'
	| 'disconnected'
	| 'error';

/** Raw kline array from GET /api/v3/klines */
type BinanceKlineRow = [
	number,
	string,
	string,
	string,
	string,
	string,
	number,
	string,
	number,
	string,
	string,
	string,
];

export interface BinanceKlineWsPayload {
	e: 'kline';
	E: number;
	s: string;
	k: {
		t: number;
		T: number;
		s: string;
		i: string;
		o: string;
		c: string;
		h: string;
		l: string;
		v: string;
		n: number;
		x: boolean;
	};
}

function rowToChartCandle(row: BinanceKlineRow): ChartCandle {
	return {
		time: Math.floor(row[0] / 1000),
		open: Number.parseFloat(row[1]),
		high: Number.parseFloat(row[2]),
		low: Number.parseFloat(row[3]),
		close: Number.parseFloat(row[4]),
	};
}

export function wsKlineToChartCandle(k: BinanceKlineWsPayload['k']): ChartCandle {
	return {
		time: Math.floor(k.t / 1000),
		open: Number.parseFloat(k.o),
		high: Number.parseFloat(k.h),
		low: Number.parseFloat(k.l),
		close: Number.parseFloat(k.c),
	};
}

export interface BinanceTradeWsPayload {
	e: 'trade';
	p: string;
}

export function getBinanceKlineStreamUrl(
	symbol: string,
	interval: BinanceKlineInterval,
): string {
	const stream = `${symbol.toLowerCase()}@kline_${interval}`;
	return `${BINANCE_WS_SINGLE}/${stream}`;
}

export function getBinanceTradeStreamName(symbol: string): string {
	return `${symbol.toLowerCase()}@trade`;
}

/** Combined stream: kline + trades on one WebSocket. */
export function getBinanceCombinedStreamUrl(streams: string[]): string {
	const path = streams.map((s) => s.toLowerCase()).join('/');
	return `${BINANCE_WS_COMBINED}/stream?streams=${path}`;
}

/** Update the open 5m (or any TF) candle with the latest trade price. */
export function patchFormingCandleWithPrice(
	candles: ChartCandle[],
	price: number,
): ChartCandle[] | null {
	if (candles.length === 0 || !Number.isFinite(price)) {
		return null;
	}

	const last = candles[candles.length - 1];
	const updated: ChartCandle = {
		...last,
		close: price,
		high: Math.max(last.high, price),
		low: Math.min(last.low, price),
	};

	if (
		updated.close === last.close &&
		updated.high === last.high &&
		updated.low === last.low
	) {
		return null;
	}

	return candles.slice(0, -1).concat(updated);
}

const BINANCE_KLINES_PAGE_SIZE = 1000;

async function fetchBinanceKlinesPage(
	symbol: string,
	interval: BinanceKlineInterval,
	limit: number,
	endTimeMs?: number,
): Promise<ChartCandle[]> {
	const params = new URLSearchParams({
		symbol: symbol.toUpperCase(),
		interval,
		limit: String(Math.min(BINANCE_KLINES_PAGE_SIZE, Math.max(1, limit))),
	});
	if (endTimeMs != null) {
		params.set('endTime', String(endTimeMs));
	}

	const response = await fetch(
		`${BINANCE_REST_BASE}/api/v3/klines?${params.toString()}`,
	);
	if (!response.ok) {
		const body = await response.text().catch(() => '');
		throw new Error(
			`Binance klines failed (${response.status})${body ? `: ${body}` : ''}`,
		);
	}

	const rows = (await response.json()) as BinanceKlineRow[];
	return rows.map(rowToChartCandle);
}

/** Loads up to `limit` most recent candles (paginates when limit > 1000). */
export async function fetchBinanceKlines(
	symbol: string,
	interval: BinanceKlineInterval,
	limit: number,
): Promise<ChartCandle[]> {
	const target = Math.max(1, limit);
	const merged: ChartCandle[] = [];
	let endTimeMs: number | undefined;

	while (merged.length < target) {
		const batchLimit = Math.min(
			BINANCE_KLINES_PAGE_SIZE,
			target - merged.length,
		);
		const batch = await fetchBinanceKlinesPage(
			symbol,
			interval,
			batchLimit,
			endTimeMs,
		);
		if (batch.length === 0) {
			break;
		}

		merged.unshift(...batch);
		endTimeMs = batch[0].time * 1000 - 1;
		if (batch.length < batchLimit) {
			break;
		}
	}

	return merged.length > target ? merged.slice(-target) : merged;
}

export function mergeKlineUpdate(
	candles: ChartCandle[],
	update: ChartCandle,
	maxCandles: number,
): ChartCandle[] {
	if (candles.length === 0) {
		return [update];
	}

	const last = candles[candles.length - 1];
	if (update.time === last.time) {
		const next = candles.slice(0, -1).concat(update);
		return next.length > maxCandles ? next.slice(-maxCandles) : next;
	}

	if (update.time > last.time) {
		const next = candles.concat(update);
		return next.length > maxCandles ? next.slice(-maxCandles) : next;
	}

	const index = candles.findIndex((c) => c.time === update.time);
	if (index === -1) {
		return candles;
	}
	const next = candles.slice();
	next[index] = update;
	return next;
}

File: src/PolyTrader.Infrastructure/Binance/BinanceMarketService.cs
using System.Globalization;
using System.Net.Http.Json;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using PolyTrader.Core.Models;
using PolyTrader.Infrastructure.Options;

namespace PolyTrader.Infrastructure.Binance;

public sealed class BinanceKlineClosedEventArgs : EventArgs
{
    public required ChartCandle Candle { get; init; }
    public required IReadOnlyList<ChartCandle> Buffer { get; init; }
}

public interface IBinanceMarketService
{
    BinanceConnectionStatus Status { get; }
    IReadOnlyList<ChartCandle> Candles { get; }
    double? LastPrice { get; }
    event EventHandler<BinanceKlineClosedEventArgs>? KlineClosed;
    event EventHandler? CandlesUpdated;
    Task StartAsync(CancellationToken cancellationToken = default);
    Task StopAsync();
    /// <summary>
    /// Overwrite recent bars in the in-memory buffer with Binance REST OHLC (authoritative closed values).
    /// </summary>
    Task RefreshRecentCandlesAsync(int limit = 100, CancellationToken cancellationToken = default);
}

public enum BinanceConnectionStatus
{
    Idle,
    Loading,
    Connected,
    Reconnecting,
    Disconnected,
    Error
}

public sealed class BinanceMarketService : IBinanceMarketService, IAsyncDisposable
{
    private const string RestBase = "https://api.binance.com";
    private const string WsSingle = "wss://stream.binance.com:9443/ws";
    private const int MaxReconnectDelayMs = 30_000;

    private readonly PolyTraderOptions _options;
    private readonly IHttpClientFactory _httpClientFactory;
    private readonly ILogger<BinanceMarketService> _logger;
    private readonly object _lock = new();
    private readonly List<ChartCandle> _candles = [];

    private ClientWebSocket? _ws;
    private CancellationTokenSource? _cts;
    private Task? _receiveTask;
    private int _reconnectAttempt;

    public BinanceMarketService(
        IOptions<PolyTraderOptions> options,
        IHttpClientFactory httpClientFactory,
        ILogger<BinanceMarketService> logger)
    {
        _options = options.Value;
        _httpClientFactory = httpClientFactory;
        _logger = logger;
    }

    public BinanceConnectionStatus Status { get; private set; } = BinanceConnectionStatus.Idle;
    public IReadOnlyList<ChartCandle> Candles
    {
        get { lock (_lock) return _candles.ToList(); }
    }

    public double? LastPrice { get; private set; }
    public event EventHandler<BinanceKlineClosedEventArgs>? KlineClosed;
    public event EventHandler? CandlesUpdated;

    public async Task StartAsync(CancellationToken cancellationToken = default)
    {
        await StopAsync();
        _logger.LogInformation(
            "Binance market service starting symbol={Symbol} interval={Interval}",
            _options.BinanceSymbol,
            _options.BinanceInterval);
        _cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        Status = BinanceConnectionStatus.Loading;
        await LoadHistoryAsync(_cts.Token);
        _logger.LogInformation("Binance history loaded: {Count} candles", Candles.Count);
        _ = RunWebSocketLoopAsync(_cts.Token);
    }

    public async Task RefreshRecentCandlesAsync(int limit = 100, CancellationToken cancellationToken = default)
    {
        limit = Math.Clamp(limit, 1, 1000);
        var client = _httpClientFactory.CreateClient();
        var symbol = _options.BinanceSymbol.ToUpperInvariant();
        var url =
            $"{RestBase}/api/v3/klines?symbol={symbol}&interval={_options.BinanceInterval}&limit={limit}";
        var rows = await client.GetFromJsonAsync<JsonElement[]>(url, cancellationToken);
        if (rows is not { Length: > 0 })
        {
            return;
        }

        var fresh = rows.Select(ParseKlineRow).ToList();
        lock (_lock)
        {
            foreach (var candle in fresh)
            {
                var idx = _candles.FindIndex(c => c.Time == candle.Time);
                if (idx >= 0)
                {
                    _candles[idx] = candle;
                }
                else
                {
                    _candles.Add(candle);
                }
            }

            _candles.Sort((a, b) => a.Time.CompareTo(b.Time));
            while (_candles.Count > _options.CandleHistoryLimit)
            {
                _candles.RemoveAt(0);
            }
        }

        CandlesUpdated?.Invoke(this, EventArgs.Empty);
    }

    public async Task StopAsync()
    {
        _cts?.Cancel();
        if (_ws is { State: WebSocketState.Open })
        {
            try { await _ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "stop", CancellationToken.None); }
            catch { /* ignore */ }
        }

        _ws?.Dispose();
        _ws = null;
        if (_receiveTask != null)
        {
            try { await _receiveTask; } catch { /* ignore */ }
        }

        Status = BinanceConnectionStatus.Disconnected;
        _logger.LogInformation("Binance market service stopped");
    }

    private async Task LoadHistoryAsync(CancellationToken ct)
    {
        const int pageSize = 1000;
        var client = _httpClientFactory.CreateClient();
        var symbol = _options.BinanceSymbol.ToUpperInvariant();
        var target = Math.Max(1, _options.CandleHistoryLimit);
        var merged = new List<ChartCandle>();
        long? endTimeMs = null;

        while (merged.Count < target)
        {
            var batchLimit = Math.Min(pageSize, target - merged.Count);
            var url =
                $"{RestBase}/api/v3/klines?symbol={symbol}&interval={_options.BinanceInterval}&limit={batchLimit}";
            if (endTimeMs.HasValue)
            {
                url += $"&endTime={endTimeMs.Value}";
            }

            var rows = await client.GetFromJsonAsync<JsonElement[]>(url, ct);
            if (rows is not { Length: > 0 })
            {
                break;
            }

            var batch = rows.Select(ParseKlineRow).ToList();
            merged.InsertRange(0, batch);
            endTimeMs = batch[0].Time * 1000 - 1;
            if (rows.Length < batchLimit)
            {
                break;
            }
        }

        if (merged.Count > target)
        {
            merged = merged.Skip(merged.Count - target).ToList();
        }

        if (merged.Count == 0)
        {
            throw new InvalidOperationException("Empty kline history");
        }

        lock (_lock)
        {
            _candles.Clear();
            _candles.AddRange(merged);
        }

        CandlesUpdated?.Invoke(this, EventArgs.Empty);
    }

    private async Task RunWebSocketLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                await ConnectAndReceiveAsync(ct);
                _reconnectAttempt = 0;
            }
            catch (OperationCanceledException) when (ct.IsCancellationRequested)
            {
                break;
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Binance WS error");
                Status = BinanceConnectionStatus.Error;
            }

            if (ct.IsCancellationRequested) break;

            Status = BinanceConnectionStatus.Reconnecting;
            var delay = Math.Min(MaxReconnectDelayMs, (int)Math.Pow(2, _reconnectAttempt) * 1000);
            _reconnectAttempt++;
            await Task.Delay(delay, ct);
        }
    }

    private async Task ConnectAndReceiveAsync(CancellationToken ct)
    {
        _ws?.Dispose();
        _ws = new ClientWebSocket();
        var symbol = _options.BinanceSymbol.ToLowerInvariant();
        var stream = $"{symbol}@kline_{_options.BinanceInterval}";
        var uri = new Uri($"{WsSingle}/{stream}");
        await _ws.ConnectAsync(uri, ct);
        Status = BinanceConnectionStatus.Connected;
        _logger.LogInformation("Binance WebSocket connected stream={Stream}", stream);

        var buffer = new byte[8192];
        while (_ws.State == WebSocketState.Open && !ct.IsCancellationRequested)
        {
            var result = await _ws.ReceiveAsync(buffer, ct);
            if (result.MessageType == WebSocketMessageType.Close) break;

            var json = Encoding.UTF8.GetString(buffer, 0, result.Count);
            ProcessMessage(json);
        }
    }

    private void ProcessMessage(string json)
    {
        using var doc = JsonDocument.Parse(json);
        if (!doc.RootElement.TryGetProperty("k", out var k)) return;

        var candle = new ChartCandle
        {
            Time = k.GetProperty("t").GetInt64() / 1000,
            Open = double.Parse(k.GetProperty("o").GetString()!, CultureInfo.InvariantCulture),
            High = double.Parse(k.GetProperty("h").GetString()!, CultureInfo.InvariantCulture),
            Low = double.Parse(k.GetProperty("l").GetString()!, CultureInfo.InvariantCulture),
            Close = double.Parse(k.GetProperty("c").GetString()!, CultureInfo.InvariantCulture)
        };

        LastPrice = candle.Close;
        var isClosed = k.GetProperty("x").GetBoolean();

        IReadOnlyList<ChartCandle> snapshot;
        lock (_lock)
        {
            var idx = _candles.FindIndex(c => c.Time == candle.Time);
            if (idx >= 0) _candles[idx] = candle;
            else _candles.Add(candle);

            _candles.Sort((a, b) => a.Time.CompareTo(b.Time));
            while (_candles.Count > _options.CandleHistoryLimit)
            {
                _candles.RemoveAt(0);
            }

            snapshot = _candles.ToList();
        }

        CandlesUpdated?.Invoke(this, EventArgs.Empty);

        if (isClosed)
        {
            _logger.LogInformation(
                "Binance kline closed {Time} O={Open} H={High} L={Low} C={Close}",
                candle.Time,
                candle.Open,
                candle.High,
                candle.Low,
                candle.Close);
            KlineClosed?.Invoke(this, new BinanceKlineClosedEventArgs
            {
                Candle = candle,
                Buffer = snapshot
            });
        }
    }

    private static ChartCandle ParseKlineRow(JsonElement row)
    {
        return new ChartCandle
        {
            Time = row[0].GetInt64() / 1000,
            Open = double.Parse(row[1].GetString()!, CultureInfo.InvariantCulture),
            High = double.Parse(row[2].GetString()!, CultureInfo.InvariantCulture),
            Low = double.Parse(row[3].GetString()!, CultureInfo.InvariantCulture),
            Close = double.Parse(row[4].GetString()!, CultureInfo.InvariantCulture)
        };
    }

    public async ValueTask DisposeAsync() => await StopAsync();
}

internal sealed class BinanceCombinedMessage
{
    [JsonPropertyName("stream")]
    public string? Stream { get; set; }

    [JsonPropertyName("data")]
    public JsonElement Data { get; set; }
}
